eth/backend.go

```diff
@@ -0,0 +1,171 @@
+package eth
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"strings"
+
+	ethereum "github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	ethabi "github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/golang/glog"
+	"github.com/livepeer/go-livepeer/eth/contracts"
+)
+
+var abis = []string{
+	contracts.BondingManagerABI,
+	contracts.ControllerABI,
+	contracts.LivepeerTokenABI,
+	contracts.LivepeerTokenFaucetABI,
+	contracts.MinterABI,
+	contracts.RoundsManagerABI,
+	contracts.ServiceRegistryABI,
+	contracts.TicketBrokerABI,
+}
+
+type Backend interface {
+	ethereum.ChainReader
+	ethereum.ChainStateReader
+	ethereum.ChainSyncReader
+	ethereum.TransactionReader
+	ethereum.TransactionSender
+	ethereum.ContractCaller
+	ethereum.PendingContractCaller
+	ethereum.PendingStateReader
+	// ethereum.PendingStateEventer
+	ethereum.GasEstimator
+	ethereum.GasPricer
+	ethereum.LogFilterer
+
+	NetworkID(ctx context.Context) (*big.Int, error)
+}
+
+type backend struct {
+	*ethclient.Client
+	methods      map[string]*method
+	nonceManager *NonceManager
+}
+
+type method struct {
+	name string
+	abi  *abi.ABI
+}
+
+func NewBackend(client *ethclient.Client) (Backend, error) {
+	methods, err := makeMethodsMap()
+	if err != nil {
+		return nil, err
+	}
+
+	return &backend{
+		client,
+		methods,
+		NewNonceManager(client),
+	}, nil
+}
+
+func (b *backend) PendingNonceAt(ctx context.Context, account common.Address) (uint64, error) {
+	var nonce uint64
+	var err error
+	if b.nonceManager != nil {
+		b.nonceManager.Lock(account)
+		defer func() {
+			if err == nil {
+				b.nonceManager.Update(account, nonce)
+			}
+
+			b.nonceManager.Unlock(account)
+		}()
+
+		nonce, err = b.nonceManager.Next(account)
+		if err != nil {
+			return 0, err
+		}
+	} else {
+		nonce, err = b.Client.PendingNonceAt(ctx, account)
+		if err != nil {
+			return 0, fmt.Errorf("failed to retrieve account nonce: %v", err)
+		}
+	}
+
+	return nonce, nil
+}
+
+func (b *backend) SendTransaction(ctx context.Context, tx *types.Transaction) error {
+	err := b.Client.SendTransaction(ctx, tx)
+
+	// decode tx data
+	data := tx.Data()
```

Comment:
I added error checking, ran this locally and I'm seeing a bunch of errors such as:

```
E1024 12:07:58.759626   57619 backend.go:113] abi: improperly encoded boolean value
```

```
E1024 12:07:04.060205   57594 backend.go:113] abi: unmarshalling empty output
```